upload-api/ucan-invocation.js

```diff
@@ -57,6 +55,80 @@ export async function parseUcanInvocationRequest(request) {
   }
 }
 
+/**
+ * @param {UcanInvocationWrapper} ucanInvocation
+ * @param {string} ucanLogStreamName
+ */
+export function getKinesisRecord (ucanInvocation, ucanLogStreamName) {
+  return {
+    Data: uint8arrayFromString(
+      JSON.stringify({
+        carCid: ucanInvocation.carCid,
+        value: ucanInvocation.value,
+        ts: Date.now(),
+      })
+    ),
+    // https://docs.aws.amazon.com/streams/latest/dev/key-concepts.html
+    // A partition key is used to group data by shard within a stream.
+    // It is required, and now we are starting with one shard. We need to study best partition key
+    PartitionKey: 'key',
+    StreamName: ucanLogStreamName,
+  }
+}
+
+/**
+ * @typedef {object} UcanInvocationCtx
+ * @property {import('@web3-storage/upload-api').UcanBucket} storeBucket
+ * @property {string} basicAuth
+ * @property {string} [streamName]
+ * @property {import('@aws-sdk/client-kinesis').Kinesis} [kinesisClient]
+ */
+
+/**
+ * @param {import('aws-lambda').APIGatewayProxyEventV2} request
+ * @param {UcanInvocationCtx} ctx
+ */
+export async function processUcanInvocation(request, ctx) {
+  const token = getTokenFromRequest(request)
+  if (token !== ctx.basicAuth) {
+    throw new NoValidTokenError('invalid Authorization credentials provided')
+  }
+
+  const ucanInvocation = await parseUcanInvocationRequest(request)
+
+  // persist successful invocation
+  await persistUcanInvocation(ucanInvocation, ctx.storeBucket)
+
+  // Put invocation to UCAN stream
+  ctx.streamName && await ctx.kinesisClient?.putRecord(
+    getKinesisRecord(ucanInvocation, ctx.streamName)
+  )
+}
+
+/**
+ * @param {import('aws-lambda').APIGatewayProxyEventV2} request
+ */
+function getTokenFromRequest (request) {
+  const authHeader = request.headers.Authorization || ''
+  if (!authHeader) {
+    throw new NoTokenError('no Authorization header provided')
```

Comment:
I wonder if introducing authorization token is even worth it. We allow anyone to send us requests anyway and we only reject requests if UCAN audience isn't our service (we may even lift that in the future).

In other words we'll be forwarding CAR requests here anyway so it is unclear to me what kind of abuse would auth header here prevent or why would any actor choose to send them this way as opposed to main API endpoint which would forward them here anyhow.

Only abuse I can think of is populating CAR with some garbage, but probably auth token is not going to help with that anyway. Maybe better guard here would be to limit CAR size and check that audience is our service instead ?

To be clear I'm not opposing having auth header, it's just in this instance I fail to recognize a merit.

Reply:
> We allow anyone to send us requests anyway and we only reject requests if UCAN audience isn't our service (we may even lift that in the future).

This is true. But if we receive such requests we have guarantee that they were really processed. Having this open has the opposite effect where attacker could send us Invocations that were not handled. For instance, `store/remove` to decrease used size...

Services we want to enable signing something would solve this issue. But I would avoid complexity here given this is a temporary solution until we get to work on invocation spec and make this transparent for Providers from ucanto.

Reply:
@Gozala aligned with Alan to move this way for now in sync A. This is a temporary solution anyway given we want to make this more integrated with ucanto